### PR TITLE
Feature/gui memory multiplebytes

### LINKED ISSUE #147: memory different number of bytes per cell

### PR DESCRIPTION
**Pull request by andizimmerer**
closes #147

### CODE REVIEW

**Comment by janschopohl** [comment]
Switching between a different number of bits to display seems to be broken. I think the 16bit and 32bit modes don't load the memory data from the correct address.
screenshot_memory

File: include/ui/memory-component-presenter.hpp [hunk 1]
```diff
@@ -98,6 +99,15 @@ class MemoryComponentPresenter : public QAbstractListModel {
    */
   QHash<int, QByteArray> roleNames() const Q_DECL_OVERRIDE;
 
+  /**
+   * returns the number of bytes to be displayed in one memory cell.
+   * the role is an indicator of how many bytes are chosen by the user.
+   *
+   * /param role the display role for this cell in the QAbstractListModel
+   * /return the number of bytes to be shown in one cell
```
Comment (goldsborough) [inline include/ui/memory-component-presenter.hpp:106]: lol

File: include/ui/memory-component-presenter.hpp [hunk 2]
```diff
@@ -98,6 +99,15 @@ class MemoryComponentPresenter : public QAbstractListModel {
    */
   QHash<int, QByteArray> roleNames() const Q_DECL_OVERRIDE;
 
+  /**
+   * returns the number of bytes to be displayed in one memory cell.
+   * the role is an indicator of how many bytes are chosen by the user.
+   *
+   * /param role the display role for this cell in the QAbstractListModel
```
Comment (goldsborough) [inline include/ui/memory-component-presenter.hpp:105]: `\param` and then the first letter after the variable name should be capitalized (it's the start of a sentence):
`\param role The`

File: include/ui/memory-component-presenter.hpp [hunk 3]
```diff
@@ -98,6 +99,15 @@ class MemoryComponentPresenter : public QAbstractListModel {
    */
   QHash<int, QByteArray> roleNames() const Q_DECL_OVERRIDE;
 
+  /**
+   * returns the number of bytes to be displayed in one memory cell.
```
Comment (goldsborough) [inline include/ui/memory-component-presenter.hpp]: The `\brief` part (`\brief`) is implicit here should fit on a single line. You can put more detailed description in the )(implicit) `\detail` part (after a newline). Also, start the sentence with a capital character

File: include/ui/memory-representations.hpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+class MemoryRepresentations : public QObject
+{
+    Q_OBJECT
+
+
+    public:
+    /**
+     * the number format displayed in memory
+     */
+    enum Representation
```
Comment (goldsborough) [inline include/ui/memory-representations.hpp]: this is not clang formatted

File: include/ui/memory-representations.hpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#include <QMetaType>
+
+class MemoryRepresentations : public QObject
+{
+    Q_OBJECT
+
+
+    public:
+    /**
+     * the number format displayed in memory
```
Comment (goldsborough) [inline include/ui/memory-representations.hpp]: Sentences start with a capital letter

Comment (goldsborough) [reply include/ui/memory-representations.hpp]: And end with a dot

File: include/ui/memory-representations.hpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
+#ifndef ERAGPSIM_UI_MEMORY_REPRESENTATIONS_H
+#define ERAGPSIM_UI_MEMORY_REPRESENTATIONS_H
+
+#include <QObject>
+#include <QtQml>
+#include <QMetaType>
+
+class MemoryRepresentations : public QObject
```
Comment (goldsborough) [inline include/ui/memory-representations.hpp]: class level documentaiton please -- what does this class do?

File: include/ui/memory-representations.hpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+     * the number format displayed in memory
+     */
+    enum Representation
+    {
+       BIN,
+       OCT,
+       HEX,
+       SDEC,
+       DEC
+    };
```
Comment (goldsborough) [inline include/ui/memory-representations.hpp]: newline (or clang format)

File: include/ui/memory-representations.hpp [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    Q_ENUMS(Representation)
+
+        static void declareQML() {
+            //qmlRegisterType<MemoryRepresentations>("MemoryRepresentation", 1, 0, "Representation");
```
Comment (goldsborough) [inline include/ui/memory-representations.hpp]: No comments in production code. Delete this or uncomment

File: source/ui/Components/Memory/MemoryComponent.qml [hunk 9]
```diff
@@ -20,114 +20,236 @@
...
+
+                Connections {
+                    target: tableView
+                    onColumnCountChanged: {
+                        // dynamically add columns that were added by the user
+                        while(headerDropdownList.count < tableView.columnCount - 1)
+                            headerDropdownList.append(ListElement);
+                    }
+                }
+
```
Comment (goldsborough) [inline source/ui/Components/Memory/MemoryComponent.qml:129]: so much space, i feel so free

File: source/ui/memory-component-presenter.cpp [hunk 10]
```diff
@@ -62,40 +94,77 @@ int MemoryComponentPresenter::rowCount(const QModelIndex &parent) const {
...
-    }
-    default: {
-      // unknown column role, return empty value
-      return QVariant();
-    }
+  // get role as a string because there is more information in it
+  QString role_string = MemoryComponentPresenter::roleNames().value(role);
+
+  int number_of_bits = 8;
+  if (role_string.endsWith("16")) number_of_bits = 16;
```
Comment (goldsborough) [inline source/ui/memory-component-presenter.cpp:107]: I don't think this is clang fomatted

File: source/ui/memory-component-presenter.cpp [hunk 11]
```diff
@@ -62,40 +94,77 @@ int MemoryComponentPresenter::rowCount(const QModelIndex &parent) const {
...
   }
+
+  if (role == InfoRole) {
+    // TODO fetch value from core
+    return QString("");
+  }
+
+  MemoryValue memory_cell =
+      _memoryAccess
+          .getMemoryValueAt(index.row() * (number_of_bits / 8),
```
Comment (goldsborough) [inline source/ui/memory-component-presenter.cpp]: Please use a temporary variable, this looks too artistic

File: source/ui/memory-component-presenter.cpp [hunk 12]
```diff
@@ -62,40 +94,77 @@ int MemoryComponentPresenter::rowCount(const QModelIndex &parent) const {
...
+                            number_of_bits / 8)
+          .get();
+  std::string stringvalue;
```
Comment (goldsborough) [inline source/ui/memory-component-presenter.cpp]: `stringValue` or just `string`

File: source/ui/memory-component-presenter.cpp [hunk 13]
```diff
@@ -62,40 +94,77 @@ int MemoryComponentPresenter::rowCount(const QModelIndex &parent) const {
...
+  if (role_string.startsWith("bin"))
```
Comment (goldsborough) [inline source/ui/memory-component-presenter.cpp]: nein, nein, nein, braces.

The reason is that some Kleinpraktikant might come along and think, hm, I'll add another statement under this `if`:
```
if (role_string.startsWith("bin"))
std::clog << "starts with bin" << std::endl;
stringvalue = StringConversions::toBinString(memory_cell);
```

Question to the audience: what's wrong here?

**Review by goldsborough — APPROVED** [review]

**Comment by andizimmerer** [comment]
Thank you for your review! `memory-representations.hpp` was an unused file which was meant for replacing string comparisons with enums. But this functionality was dropped.

File: source/ui/Components/Memory/MemoryComponent.qml [hunk 14]
```diff
@@ -20,114 +20,236 @@
...
+                        onCurrentIndexChanged: {
+                            // depending on the usage there are 3 different actions
+                            // 1. update the number of bits in each memory cell
+                            if(model === modelBits) {
+                                number_bits = model.get(bitChooser.currentIndex).bits;
+                            }
+                            else {
+                                // 2. dynamically remove the column
+                                if(bitChooser.currentText == "remove...") {
```
Comment (FelixOpolka) [inline source/ui/Components/Memory/MemoryComponent.qml]: ===

File: source/ui/Components/Memory/MemoryComponent.qml [hunk 15]
```diff
@@ -20,114 +20,236 @@
...
-            width: parent.width - 200
+            resizable: true
+            width: 100
         }
+
         model: memoryModel
+
+        // add a column with the content for each cell at startup
+        Component.onCompleted: {
+            tableView.insertColumn( tableView.columnCount - 1, column);
```
Comment (FelixOpolka) [inline source/ui/Components/Memory/MemoryComponent.qml]: Extraneous space.

File: source/ui/Components/Memory/MemoryComponent.qml [hunk 16]
```diff
@@ -20,114 +20,236 @@
...
+                // move header left and right as the tableView is moved by the horizontal scroll bars
+                x: -tableView.flickableItem.contentX
+
+                orientation: Qt.Horizontal
+
+                Connections {
+                    target: tableView
+                    onColumnCountChanged: {
+                        // dynamically add columns that were added by the user
+                        while(headerDropdownList.count < tableView.columnCount - 1)
```
Comment (FelixOpolka) [inline source/ui/Components/Memory/MemoryComponent.qml]: Missing { ... }

**Review by FelixOpolka — APPROVED** [review]
Looks really good!

**Pull request merged**